numpy/random/mtrand/mtrand.pyx

```diff
@@ -4640,6 +4642,368 @@ cdef class RandomState:
+        rk_hypergeometric:
+
+            green_sample = rk_hypergeometric(ngood=ngreen,
+                                             nbad=total - nred - ngreen,
+                                             nsample=nsample - red_sample)
+
+        Similarly,
+
+            blue_sample = rk_hypergeometric(
+                              ngood=nblue,
+                              nbad=total - nred - ngreen - nblue,
+                              nsample=nsample - red_sample - green_sample)
+
+        Finally,
+
+            yellow_sample = total - (red_sample + green_sample + blue_sample).
+
+        The above sequence of steps is implemented as a loop for an arbitrary
+        number of colors in the innermost loop in the code below.  `remaining`
+        is the value passed to `nbad`; it is `total - colors[0]` in the first
+        call to rk_hypergeometric(), and then decreases by `colors[j]` in
+        each iteration.  `num_to_sample` is the `nsample` argument.  It
+        starts at this function's `nsample` input, and is decreased by the
+        result of the call to rk_hypergeometric() in each iteration.
+        """
+
+        cdef npy_intp total, ncolors, i, j, num_to_sample, sz, remaining, r
+        cdef int more_than_half
+        cdef long *colors_ptr
+        cdef ndarray samples
+        cdef long *samples_ptr
+
+        colors_ptr = <long*> PyArray_DATA(colors)
+        total = np.sum(colors)
+
+        ncolors = len(colors)
+        shape = _shape_from_size(size, ncolors)
+        samples = np.zeros(shape, dtype=np.int_)
+        samples_ptr = <long*> PyArray_DATA(samples)
+        sz = PyArray_SIZE(samples)
+
+        more_than_half = nsample > (total // 2)
+        if more_than_half:
+            nsample = total - nsample
+
+        with self.lock, nogil:
+            i = 0
+            while i < sz:
+                num_to_sample = nsample
+                remaining = total
+                for j in range(ncolors - 1):
+                    if num_to_sample <= 0:
+                        break
+                    remaining -= colors_ptr[j]
+                    r = rk_hypergeometric(self.internal_state, colors_ptr[j],
+                                          remaining, num_to_sample)
+                    samples_ptr[i+j] = r
+                    num_to_sample -= r
+                else:
+                    samples_ptr[i + ncolors - 1] = num_to_sample
+
+                if more_than_half:
+                    for j in range(ncolors):
+                        samples_ptr[i+j] = colors_ptr[j] - samples_ptr[i+j]
+
+                i += ncolors
+
+        return samples
+
+    def _multivar_hypergeom_count(self, ndarray colors, npy_intp nsample,
+                                  size=None):
+        """
+        Draw samples from the multivariate hypergeometric distribution
+        ("count" algorithm).
+
+        `colors` must be a one-dimensional C-contiguous numpy array
```

Comment:
True.  When I was trying to implement the automatic selection, it was convenient to keep the algorithms as accessible methods so I could bypass the automatic part and test the methods individually.  Now that there is a "method" argument, there is nothing lost by moving the two algorithms to functions outside of the `RandomState` class.  (Yeah, I realize we're talking about a suggestion that you made nine months ago. :)